ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp

```diff
@@ -1065,12 +1070,13 @@ Tensor matmul(
                 create_matmul_struct(input_tensor_a, input_tensor_b, parameters),
                 {input_tensor_a, input_tensor_b},
                 optional_input_tensors,
-                {},
+                optional_output_tensors,
                 queue_id);
         },
         {input_tensor_a, input_tensor_b},
         output_tensors,
-        optional_input_tensors);
+        optional_input_tensors,
+        {optional_output_tensor});
```

Comment:
You need to validate on optional output tensors as well.

See e.g. ttnn/cpp/ttnn/operation.hpp
```
template <class T>
constexpr bool implements_validate_with_output_tensors_and_optional_input_tensors() {
    return std::experimental::is_detected_v<
        has_validate_with_output_tensors_t,
        T,
        const Tensors&,                                   // input_tensors
        const std::vector<std::optional<const Tensor>>&,  // optional input_tensors
        const OptionalTensors&>;                          // optional output_tensors
}
```

It's probably too complicated too try to do what the new op framework (called TMP) does https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html